### PR TITLE
Add new rules to PMD - Optimizations

### PR DESCRIPTION
**Pull request by kanghj**
I'll do the remaining ones in separate pull requests

### CODE REVIEW

**Comment by wkurniawan07** [comment]
The `SimplifyStartsWith` rule looks kind of strange, though. Should we include it?

**Comment by kanghj** [comment]
I'll remove it. I don't like it too

**Comment by wkurniawan07** [comment]
There's a merge conflict here...

File: src/main/java/teammates/ui/controller/LoginFilter.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package teammates.ui.controller;
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/LoginFilter.java]: You just restored this class to life... :p

File: src/main/java/teammates/ui/controller/InstructorSearchPageAction.java [hunk 2]
```diff
@@ -240,17 +240,17 @@ private boolean isInstructorAllowedToViewComment(CommentAttributes commentAttrib
...
                     // prevent error--however, this should never happen unless there is corruption of data
                     return false;
                 }
                 
-                recipient = commentAttributes.recipients.iterator().next();
+                boolean isForSection = true;
+                String section = "None";
+                String recipient = "";
                 
+                recipient = commentAttributes.recipients.iterator().next();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/InstructorSearchPageAction.java]: I'm surprised FindBugs didn't catch this one.
```java
String recipient = commentAttributes.recipients.iterator().next();
```

File: src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java [hunk 3]
```diff
@@ -114,6 +113,7 @@ protected ActionResult execute() throws EntityDoesNotExistException {
             return createAjaxResult(data);
         }
         
+       BackDoorLogic backDoor = new BackDoorLogic();
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: CheckStyle will catch this when the rule is up. The indentation is off by one whitespace.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 4]
```diff
@@ -2065,6 +2061,12 @@ private FeedbackSessionResultsBundle getFeedbackSessionResultsForUserWithParams(
         for (FeedbackQuestionAttributes qn : allQuestions) {
             allQuestionsMap.put(qn.getId(), qn);
         }
+        
+        boolean isInSection = Boolean.parseBoolean(params.get("inSection"));
+        boolean isToSection = Boolean.parseBoolean(params.get("toSection"));
+        boolean isFromSection = Boolean.parseBoolean(params.get("fromSection"));
+        boolean isComplete = params.get("range") != null ? false : true;
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: There's a PMD rule that will catch this, but let's fix it now. `params.get("range") == null` will do the trick.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 5]
```diff
@@ -444,13 +444,14 @@ public FeedbackSessionQuestionsBundle getFeedbackSessionQuestionsForStudent(
...
-                userEmail);
-
+        
         if (fsa == null) {
             throw new EntityDoesNotExistException(
                     "Trying to get a feedback session that does not exist.");
         }
+        
+        StudentAttributes student = studentsLogic.getStudentForEmail(courseId,
+                                        userEmail);
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: This is too short to be line-broken.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 6]
```diff
@@ -402,13 +402,13 @@ public FeedbackSessionQuestionsBundle getFeedbackSessionQuestionsForStudent(
...
                 courseId, feedbackSessionName);
-        StudentAttributes student = studentsLogic.getStudentForEmail(courseId,
-                userEmail);
 
         if (fsa == null) {
             throw new EntityDoesNotExistException(
                     "Trying to get a feedback session that does not exist.");
         }
+        StudentAttributes student = studentsLogic.getStudentForEmail(courseId,
+                                                                     userEmail);
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Similarly here.

File: src/main/java/teammates/logic/core/FeedbackResponsesLogic.java [hunk 7]
```diff
@@ -814,6 +809,8 @@ private void addNewResponses(
             return viewableResponses;
         }
 
+        StudentAttributes student = studentsLogic.getStudentForEmail(question.courseId,
+                                                                     studentEmail);
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackResponsesLogic.java]: Here, too.

File: src/main/java/teammates/common/datatransfer/FeedbackNumericalScaleQuestionDetails.java [hunk 8]
```diff
@@ -367,8 +357,10 @@ private String getStudentQuestionResultsStatisticsHtml(
         
         String statsTitle = getStatsTitle(isRecipientTypeGeneral, isRecipientTypeTeam, 
                                           hasAtLeastTwoResponsesOtherThanCurrentUser(numResponses, currentUserIdentifier, hiddenRecipients));
-        
-        html = FeedbackQuestionFormTemplates.populateTemplate(
+        String templateToUse = showAvgExcludingSelf ? 
+                                        FeedbackQuestionFormTemplates.NUMSCALE_RESULT_STATS_WITH_SELF_RESPONSE : 
+                                        FeedbackQuestionFormTemplates.NUMSCALE_RESULT_STATS;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackNumericalScaleQuestionDetails.java]: Let's use the operator-in-front style.
```java
String templateToUse = condition
                     ? valueIfTrue
                     : valueIfFalse;
```

File: src/main/java/teammates/common/datatransfer/FeedbackNumericalScaleQuestionDetails.java [hunk 9]
```diff
@@ -229,19 +220,22 @@ private String getInstructorQuestionResultsStatisticsHtml(
...
         
-        html = FeedbackQuestionFormTemplates.populateTemplate(
-                templateToUse,
-                "${summaryTitle}", statsTitle,
-                "${statsFragments}", fragmentHtml.toString());
-        
+        String statsTitle = "Response Summary";
+        String templateToUse = showAvgExcludingSelf ? 
+                               FeedbackQuestionFormTemplates.NUMSCALE_RESULT_STATS_WITH_SELF_RESPONSE :
+                               FeedbackQuestionFormTemplates.NUMSCALE_RESULT_STATS;
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackNumericalScaleQuestionDetails.java]: Here as well.

**Comment by wkurniawan07** [comment]
Added some comments.

**Pull request merged**